File: 125.验证回文串.java
/*
 * @lc app=leetcode.cn id=125 lang=java
 *
 * [125] 验证回文串
 */

// @lc code=start
class Solution {
    public boolean isPalindrome(String s) {

        int i = 0, j = s.length()-1;
        while (i < j) {
            while (!Character.isLetterOrDigit(s.charAt(i)) && i < j) i++;
            while (!Character.isLetterOrDigit(s.charAt(j)) && i < j) j--;
            if (Character.isDigit(s.charAt(i)) && s.charAt(i) != s.charAt(j)) return false;
            if (Character.isLetter(s.charAt(i))) {
                if (Character.toLowerCase(s.charAt(i)) != Character.toLowerCase(s.charAt(j)))
                    return false;
            } 
            i++;
            j--;
        }
        return true;
    }
}
// @lc code=end

File: 125.验证回文串.py
#
# @lc app=leetcode.cn id=125 lang=python3
#
# [125] 验证回文串
#

# @lc code=start
class Solution:
    def isPalindrome(self, s: str) -> bool:
        n = len(s)
        i, j = 0, n-1
        while True:
            while i < n and not s[i].isalnum():
                i += 1
            while j >= 0 and not s[j].isalnum():
                j -= 1
            if i > j:
                break
            elif s[i].lower() != s[j].lower():
                return False
            i += 1
            j -= 1
        return True
# @lc code=end

